src/charm.py

```diff
@@ -53,43 +53,32 @@ def __init__(self, *args: Any) -> None:
             strip_prefix=True,
         )
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.reset_instance_action, self._reset_instance_action)
+        self.framework.observe(self.on.reset_instance_action, self._on_reset_instance_action)
 
     def _on_config_changed(self, event: ops.HookEvent) -> None:
         """Handle changed configuration.
 
         Args:
             event: Event triggering after config is changed.
         """
-        try:
-            self._charm_state = CharmState.from_charm(charm=self)
-        except CharmConfigInvalidError as exc:
-            self.model.unit.status = ops.BlockedStatus(exc.msg)
-            return
         container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
         if not container.can_connect():
             event.defer()
             self.unit.status = ops.WaitingStatus("Waiting for pebble")
             return
-        server_name_configured = self._synapse.server_name_configured(container)
-        if (
-            server_name_configured is not None
-            and server_name_configured != self._charm_state.server_name
-        ):
-            msg = (
-                f"server_name {self._charm_state.server_name} is different from the existing one"
-                f" {server_name_configured}."
-                " Please revert the config or"
-                " run the action reset-instance if you to erase the existing instance and start a"
-                " new one."
-            )
-            self.model.unit.status = ops.BlockedStatus(msg)
-            return
-        self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
         try:
+            self._charm_state = CharmState.from_charm(charm=self)
+            self._synapse = Synapse(charm_state=self._charm_state)
+            self._synapse.check_server_name(container)
+            self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
             self._synapse.execute_migrate_config(container)
```

Comment:
DO you think that this function should check for the server name as a pre-condition? That reduces the coupling between `charm.py` and the `synapse` module

Reply:
Makes sense, changed.